File: merge-tonic-lib/tonic/hydration/embedder.py
"""Embedding interfaces and deterministic fake embedder for tests."""

from __future__ import annotations

import hashlib
from dataclasses import dataclass
from typing import Protocol


class Embedder(Protocol):
    @property
    def model_id(self) -> str: ...

    def embed_documents(self, texts: list[str]) -> list[list[float]]: ...

    def embed_query(self, text: str) -> list[float]: ...


def _deterministic_embedding(text: str, dimensions: int) -> list[float]:
    digest = hashlib.sha256(text.encode("utf-8", errors="replace")).digest()
    values: list[float] = []
    for i in range(dimensions):
        a = digest[i % len(digest)]
        b = digest[(i + 7) % len(digest)]
        values.append((a + b) / 255.0 - 1.0)
    return values


@dataclass
class DeterministicFakeEmbedder:
    model_id: str = "deterministic-fake-v1"
    dimensions: int = 16

    def embed_documents(self, texts: list[str]) -> list[list[float]]:
        return [_deterministic_embedding(text, self.dimensions) for text in texts]

    def embed_query(self, text: str) -> list[float]:
        return _deterministic_embedding(text, self.dimensions)

File: packages/tonic-core/src/hydration/embedder.ts
import { createHash } from "node:crypto";

export interface Embedder {
  readonly modelId: string;
  embedDocuments(texts: string[]): Promise<number[][]>;
  embedQuery(text: string): Promise<number[]>;
}

function deterministicEmbedding(text: string, dimensions: number): number[] {
  const hash = createHash("sha256").update(text, "utf8").digest();
  const values = new Array<number>(dimensions).fill(0);
  for (let i = 0; i < dimensions; i++) {
    const a = hash[i % hash.length] ?? 0;
    const b = hash[(i + 7) % hash.length] ?? 0;
    values[i] = (a + b) / 255 - 1;
  }
  return values;
}

export class DeterministicFakeEmbedder implements Embedder {
  readonly modelId: string;
  readonly dimensions: number;

  constructor(params: { modelId?: string; dimensions?: number } = {}) {
    this.modelId = params.modelId ?? "deterministic-fake-v1";
    this.dimensions = params.dimensions ?? 16;
  }

  async embedDocuments(texts: string[]): Promise<number[][]> {
    return texts.map((text) => deterministicEmbedding(text, this.dimensions));
  }

  async embedQuery(text: string): Promise<number[]> {
    return deterministicEmbedding(text, this.dimensions);
  }
}
